### PR TITLE
Chapter 12

### PR DESCRIPTION
**Pull request by nuriman1**

### CODE REVIEW

**Comment by nuriman1** [comment]
ready

File: app/assets/stylesheets/password_resets.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/password_resets.scss]: remove empty file

Comment (nuriman1) [reply app/assets/stylesheets/password_resets.scss]: done

**Comment by nuriman1** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  def update
+    if params[:user][:password].blank? 
+      @user.errors.add(:password, "can't be empty")
+      render :edit
+    elsif @user.update user_params 
+      log_in @user
+      flash[:success] = "Password has been reset."
+      redirect_to @user
+    else
+      flash[:danger] = "Password reset unsuccessful"
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
      flash.now[:danger] = "Password reset unsuccessful"
```

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def get_user
+    @user = User.find_by email: params[:email]
+
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
```

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+   end
+
+   def valid_user
+     return if @user && @user.activated? && @user.authenticated? :reset, params[:id]
+     redirect_to root_url
+   end
+
+   def check_expiration
+     return unless @user.password_reset_expired?
+     flash[:danger] = "Password reset has expired."
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
     return unless @user.password_reset_expired?
     
     flash[:danger] = "Password reset has expired."
```

**Review by chungpv-1008 — COMMENTED** [review]
Please fix comment, then merge this p/r yourself

File: app/controllers/password_resets_controller.rb (hunk 2 above)
Comment (nuriman1) [reply app/controllers/password_resets_controller.rb]: done

File: app/controllers/password_resets_controller.rb (hunk 3 above)
Comment (nuriman1) [reply app/controllers/password_resets_controller.rb]: done

File: app/controllers/password_resets_controller.rb (hunk 4 above)
Comment (nuriman1) [reply app/controllers/password_resets_controller.rb]: done

**Pull request merged**